akvo/rsr/tests/models/test_login_logging.py

```diff
@@ -0,0 +1,87 @@
+# For additional details on the GNU license please see < http://www.gnu.org/licenses/agpl.html >.
+
+from datetime import timedelta
+
+from django.forms import ValidationError
+from django.conf import settings
+from django.test import Client
+
+from akvo.rsr.models import LoginLog
+from akvo.rsr.models.login_log import MAX_FAILED_LOGINS
+from akvo.rsr.tests.base import BaseTestCase
+
+
+class LoginLoggingTestCase(BaseTestCase):
+    """Tests for the login logging model"""
+
+    def setUp(self):
+        self.email = 'frank@example.com'
+        self.password = 'password'
+        self.user = self.create_user(self.email, self.password)
+        self.c = Client(HTTP_HOST=settings.RSR_DOMAIN)
+
+    def test_successful_login_creates_log_entry(self):
+        # When
+        self.c.login(username=self.email, password=self.password)
+
+        # Then
+        logs = LoginLog.objects.filter(email=self.email)
+        self.assertTrue(logs.exists())
+        self.assertTrue(logs.first().success)
+
+    def test_failed_login_creates_log_entry(self):
+        # When
+        with self.assertRaises(ValidationError):
+            self.c.login(username=self.email, password='')
+
+        # Then
+        logs = LoginLog.objects.filter(email=self.email)
+        self.assertTrue(logs.exists())
+        self.assertFalse(logs.first().success)
+
+    def test_password_deactivates_after_max_attempts(self):
+        # Given
+        for _ in range(MAX_FAILED_LOGINS - 1):
+            with self.assertRaises(ValidationError):
+                self.c.login(username=self.email, password='')
+
+        # When
+        with self.assertRaises(ValidationError) as assertion:
+            self.c.login(username=self.email, password='')
+
+        # Then
+        self.assertIn('Login has been disabled', assertion.exception.message)
+
+    def test_logins_post_password_deactivation_ignored(self):
+        # When
+        for _ in range(MAX_FAILED_LOGINS + 10):
+            with self.assertRaises(ValidationError):
+                self.c.login(username=self.email, password='')
+
+        with self.assertRaises(ValidationError) as assertion:
+            self.c.login(username=self.email, password=self.password)
+
+        # Then
+        self.assertIn('Login has been disabled', assertion.exception.message)
+        logs = LoginLog.objects.filter(email=self.email)
+        self.assertEqual(MAX_FAILED_LOGINS, logs.count())
+
+    def test_login_works_after_deactivation_time(self):
+        # Given
+        for _ in range(MAX_FAILED_LOGINS + 10):
+            with self.assertRaises(ValidationError):
+                self.c.login(username=self.email, password='')
+        # HACK: Set the creation time of these login attempts to yesterday
+        creation_time = LoginLog.objects.first().created_at - timedelta(days=1)
+        LoginLog.objects.update(created_at=creation_time)
+
+        # When
+        self.c.login(username=self.email, password=self.password)
```

Comment:
If I get this right, this assumes that LOGIN_DISABLE_TIME is no larger than 86400, but I think that's an OK assumption! 😜

Reply:
Nice catch. I changed this to remove the assumption, by going back by `LOGIN_DISABLE_TIME * 2` seconds